### PR TITLE
Fix #5312: Fixes Favicon fetching to fetch from Brave-Core

### LINKED ISSUE #5312: Incorrect favicons in history panel post upgrade
### Description: 
Incorrect favicons in history panel post upgrade

### Steps to Reproduce 
  1. Install 1.37 build from app store
  2. Visit a bunch of pages to add browsing history
  3. Upgrade to 1.38 build 
  4. Upon upgrade open the history page on the active tab, site favicons are all wrong
  5. Open a new link/reload the same active tab, favicons gets fixed in history 

**Actual result:** 
image

**Expected result:**
Show correct favicons when upgraded from previous version

**Reproduces how often:** [Easily reproduced, Intermittent Issue]
Easy

**Brave Version:** 
1.38 (22.4.29.16)

**Device details:** 
iPad Pro (iOS 15.5 Beta 3)

**Website problems only:**
- did you check with Brave Shields down? NA
- did you check in Safari/Firefox (WkWebView-based browsers)?  NA

### Additional Information
cc: @Uni-verse @kjozwiak @iccub

### PR DESCRIPTION
**Pull request by Brandon-T**
## Summary of Changes
- Adds favicon fetching from Brave-Core
- Adds favicon fetching for Sync, History, Bookmarks, Passwords, Widgets, Tabs, NTP, Favourites from Brave-Core
- Does not migrate current favicons to Brave-Core

This pull request fixes #5312

## Submitter Checklist:

- [x] *Unit Tests* are updated to cover new or changed functionality
- [x] User-facing strings use `NSLocalizableString()`

## Reviewer Checklist:

- [ ] Issues include necessary QA labels:
  - `QA/(Yes|No)`
  - `bug` / `enhancement`
- [ ] Necessary [security reviews](https://github.com/brave/security/issues/new/choose) have taken place.
- [ ] Adequate unit test coverage exists to prevent regressions.
- [ ] Adequate test plan exists for QA to validate (if applicable).
- [ ] Issue and pull request is assigned to a milestone (should happen at merge time).

### CODE REVIEW

File: BraveShared/Favicons/FaviconFetcher.swift [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+    ///
+    /// Usage: NTP, Favorites
+    case largeIcon
+    /// Load smaller favicons
+    ///
+    /// Usage: History, Search, Tab Tray
+    case favicon
+  }
+  
+  // For Brave-Wallet
```
Comment (iccub) [inline BraveShared/Favicons/FaviconFetcher.swift]: This comment is too vague, maybe add more to it?

File: BraveShared/Favicons/FaviconFetcher.swift [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+  /// The size requirement for the favicon
+  public enum Kind {
+    /// Load favicons marked as `apple-touch-icon`.
+    ///
+    /// Usage: NTP, Favorites
+    case largeIcon
+    /// Load smaller favicons
+    ///
+    /// Usage: History, Search, Tab Tray
+    case favicon
```
Comment (iccub) [inline BraveShared/Favicons/FaviconFetcher.swift]: would `smallIcon` make more sense here in terms of naming?

File: BraveShared/Favicons/FaviconRenderer.swift [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+    }
+    
+    task?.cancel()
+    task = DispatchWorkItem { [weak self] in
+      guard let self = self, !self.isCancelled else {
+        taskCompletion(nil)
+        return
+      }
+      
+      // TODO: Use NSCache to cache the `attributes`
```
Comment (iccub) [inline BraveShared/Favicons/FaviconRenderer.swift]: Is this still relevant?

File: BraveShared/Favicons/FaviconRenderer.swift [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+                                                          minSizeInPoints: .init(rawValue: 0),
+                                                          fallbackToGoogleServer: false) { [weak self] attributes in
+        
+        guard let self = self, !self.isCancelled else {
+          taskCompletion(nil)
+          return
+        }
+        
+        if attributes.usesDefaultImage {
+          // taskCompletion(nil)
```
Comment (iccub) [inline BraveShared/Favicons/FaviconRenderer.swift]: what about this comment?

File: BraveShared/Favicons/UIImage+FaviconRenderer.swift [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+import Foundation
+import UIKit
+
+private struct UIConstants {
+  /// Used as backgrounds for favicons
+  static let defaultColorStrings = ["2e761a", "399320", "40a624", "57bd35", "70cf5b", "90e07f", "b1eea5", "881606", "aa1b08", "c21f09", "d92215", "ee4b36", "f67964", "ffa792", "025295", "0568ba", "0675d3", "0996f8", "2ea3ff", "61b4ff", "95cdff", "00736f", "01908b", "01a39d", "01bdad", "27d9d2", "58e7e6", "89f4f5", "c84510", "e35b0f", "f77100", "ff9216", "ffad2e", "ffc446", "ffdf81", "911a2e", "b7223b", "cf2743", "ea385e", "fa526e", "ff7a8d", "ffa7b3"]
```
Comment (iccub) [inline BraveShared/Favicons/UIImage+FaviconRenderer.swift]: line too long, please split

File: Client/Frontend/Browser/Helpers/FaviconHelper.swift [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+  }
+
+  func scriptMessageHandlerName() -> String? {
+    return "FaviconUrlsHandler"
+  }
+
+  func userContentController(_ userContentController: WKUserContentController, didReceiveScriptMessage message: WKScriptMessage, replyHandler: (Any?, String?) -> Void) {
+    defer { replyHandler(nil, nil) }
+    
+    if !InternalURL.isValid(url: message.webView!.url!),
```
Comment (iccub) [inline Client/Frontend/Browser/Helpers/FaviconHelper.swift]: can we unrwap `message.webView` here?

File: Client/Frontend/UserContent/UserScripts/FaviconHelper.js [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+          var favicon = {rel: links[i].rel.toLowerCase(), href: links[i].href};
+          if (links[i].sizes && links[i].sizes.value) {
+            favicon.sizes = links[i].sizes.value;
+          }
+          favicons.push(favicon);
+        }
+      }
+    }
+    
+    // DEBUGGING! DO NOT REMOVE!
```
Comment (iccub) [inline Client/Frontend/UserContent/UserScripts/FaviconHelper.js:33]: Is this minified/removed in prod builds?

**Review by iccub — CHANGES_REQUESTED** [review]

File: Client/Frontend/Settings/BraveCoreDebugSwitchesView.swift [hunk 8]
```diff
@@ -16,20 +16,12 @@ extension BraveCoreSwitch {
       return "Component Updater"
     case .syncURL:
       return "Sync URL"
-    case .skusEnvironment:
```
Comment (soner-yuksel) [inline Client/Frontend/Settings/BraveCoreDebugSwitchesView.swift]: This part should also be moved to a separate PR as reference for 
https://github.com/brave/brave-ios/issues/5643

File: Client/Frontend/Login/LoginListViewController.swift [hunk 9]
```diff
@@ -252,14 +252,8 @@ extension LoginListViewController {
         if let signOnRealmURL = URL(string: loginInfo.signOnRealm) {
           let domain = Domain.getOrCreate(forUrl: signOnRealmURL, persistent: true)
 
-          cell.imageView?.loadFavicon(
-            for: signOnRealmURL,
-            domain: domain,
-            fallbackMonogramCharacter: signOnRealmURL.baseDomain?.first,
-            shouldClearMonogramFavIcon: false,
-            cachedOnly: true)
+          cell.imageView?.loadFavicon(for: signOnRealmURL)
```
Comment (soner-yuksel) [inline Client/Frontend/Login/LoginListViewController.swift:255]: 🍾

File: Client/Frontend/Browser/ImageCache/WebImageCache.swift [hunk 10]
```diff
@@ -64,12 +64,6 @@ final public class WebImageCache: ImageCacheProtocol {
     let imageOperation = webImageManager.loadImage(with: url, options: SDWebImageOptions(webImageOptions), context: webImageContext, progress: progressBlock) { image, data, error, webImageCacheType, _, imageURL in
       let key = self.webImageManager.cacheKey(for: url)
       if let image = image {
-        let maxSize = CGFloat(FaviconMO.maxSize)
```
Comment (soner-yuksel) [inline Client/Frontend/Browser/ImageCache/WebImageCache.swift:67]: Are we still using the `load` method here ? Maybe we can remove it all together. As far as I see we were only using this for fav_icons before.

File: BraveWallet/FaviconReader.swift [hunk 11]
```diff
@@ -38,19 +18,19 @@ struct FaviconReader<Content: View>: View {
...
-        load(newValue, transaction: Transaction(), using: renderer)
+        load(newValue, transaction: Transaction())
       }
   }
   
-  private func load(_ url: URL?, transaction: Transaction, using renderer: WalletFaviconRenderer) {
+  private func load(_ url: URL?, transaction: Transaction) {
     guard let url = url else { return }
-    renderer.loadIcon(siteURL: url, persistent: true) { image in
+    self.faviconTask = FaviconFetcher.loadIcon(url: url, kind: .largeIcon, persistent: true, completion: { favicon in
```
Comment (soner-yuksel) [inline BraveWallet/FaviconReader.swift:30]: Is the reason we use `.largeIcon` here is because `renderer` `loadIcon` was loading large icons? Just trying to understand in which case we use large or small icons.

File: BraveShared/Favicons/FaviconFetcher.swift [hunk 12]
```diff
@@ -0,0 +1,129 @@
...
+    }
+    
+    public func cancel() {
+      task?.cancel()
+      task = nil
+    }
+  }
+  
+  private class FaviconOperation: Hashable, Equatable {
+    let id = UUID().uuidString
```
Comment (soner-yuksel) [inline BraveShared/Favicons/FaviconFetcher.swift]: I am little bit confused here. Why do we need a set of favicon operations when all the equatable identifier of these operations are same (UUID().uuidString). Cant we own only one operation and not add / remove it.

**Review by soner-yuksel — COMMENTED** [review]
Just couple of questions. thanks for this conversion and UIImageView integration for it.
I will be updating core 1.41 version in a different PR very soon and we can rebase.

File: BraveWallet/FaviconReader.swift (hunk 11 above)
Comment (Brandon-T) [reply BraveWallet/FaviconReader.swift:30]: Yes it's for specifically `bundled` icons. Atm, the parameter is ignored. But if we feel the need to use `large` icons only, we have the ability to. IE: Setting the `minSize` parameter.

File: Client/Frontend/Browser/ImageCache/WebImageCache.swift (hunk 10 above)
Comment (Brandon-T) [reply Client/Frontend/Browser/ImageCache/WebImageCache.swift:67]: It is used in some `MetadataParser` file.

File: Sources/Shared/UIColorExtensions.swift [hunk 13]
```diff
@@ -37,4 +37,10 @@ extension UIColor {
     let colorInt = UInt32(Scanner(string: string).scanInt32(representation: .hexadecimal) ?? 0)
     self.init(rgb: (Int)(colorInt))
   }
+  
+  public var rgb: Int {
+    var (r, g, b): (CGFloat, CGFloat, CGFloat) = (0.0, 0.0, 0.0)
+    getRed(&r, green: &g, blue: &b, alpha: nil)
+    return (Int(r * 255.0) << 16) | (Int(g * 255.0) << 8) | Int(b * 255.0)
+  }
```
Comment (kylehickinson) [inline Sources/Shared/UIColorExtensions.swift:45]: Please move this to `BraveUI`, we are no longer adding code to `Shared`

File: BraveShared/Extensions/FileManagerExtension.swift [hunk 14]
```diff
@@ -17,7 +17,7 @@ extension FileManager {
 
   /// URL where files downloaded by user are stored.
   /// If the download folder doesn't exists it creates a new one
-  func downloadsPath() throws -> URL {
+  public func downloadsPath() throws -> URL {
```
Comment (kylehickinson) [inline BraveShared/Extensions/FileManagerExtension.swift:20]: Move this extension/the functions you made public to `BraveShared`.  We shouldn't be making more things public in `Brave`, things should be moving out of the module instead

File: BraveShared/Favicons/Favicon.swift [hunk 15]
```diff
@@ -0,0 +1,48 @@
...
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+import Foundation
+import UIKit
+
+/// Structure representing a Favicon
+public class Favicon: Codable {
+  public let image: UIImage?
+  public let isMonogramImage: Bool
+  public let backgroundColor: UIColor
```
Comment (kylehickinson) [inline BraveShared/Favicons/Favicon.swift:13]: Why do both `FaviconAttributesContainer` and `Favicon` both exist now? Seems like we should only have one of them

File: BraveShared/Favicons/FaviconFetcher.swift [hunk 16]
```diff
@@ -0,0 +1,146 @@
...
+  
+  public static func clearCache() {
+    SDImageCache.shared.memoryCache.removeAllObjects()
+    SDImageCache.shared.diskCache.removeAllData()
+  }
+  
+  @discardableResult
+  public static func loadIcon(url: URL, kind: FaviconFetcher.Kind = .smallIcon, persistent: Bool, completion: ((Favicon?) -> Void)?) -> Cancellable {
+    let cancellable = Cancellable { (self, bundledRenderer, faviconRenderer) in
+      guard let self = self, !self.isCancelled else {
```
Comment (kylehickinson) [inline BraveShared/Favicons/FaviconFetcher.swift]: Can we rename `self` here? It's going to be confusing to read

File: BraveShared/Favicons/FaviconFetcher.swift [hunk 17]
```diff
@@ -0,0 +1,146 @@
...
+    return URLOrigin(url: url).url ?? url
+  }
+  
+  private static func storeInCache(_ favicon: Favicon, for url: URL, persistent: Bool) {
+    // Do not cache non-persistent icons
+    // Do not cache monogram icons
+    if !persistent && !favicon.isMonogramImage {
+      guard let data = try? JSONEncoder().encode(favicon) else {
+        return
+      }
```
Comment (kylehickinson) [inline BraveShared/Favicons/FaviconFetcher.swift]: The initial expression seems to be wrong? The comment says "Do not cache non-persistant icons" but then checking if persistent is false to cache. I assume you want "if persistent and not a monogram"

These expressions can also be merged

`guard persistant, !favicon.isMonogramImage, let data = try? JSONEncoder().encode(favicon) else { return }`

Comment (kylehickinson) [reply BraveShared/Favicons/FaviconFetcher.swift]: Also I recommend _not_ doing `try?` otherwise you can't even log why things fail. I recommend using standard `try` and do a `log.error` with the error on failure

File: BraveShared/Favicons/FaviconFetcher.swift [hunk 18]
```diff
@@ -0,0 +1,146 @@
...
+    let cachedURL = cacheURL(for: url)
+    let data = SDImageCache.shared.memoryCache.object(forKey: cachedURL.absoluteString) as? Data ??
+               SDImageCache.shared.diskCache.data(forKey: cachedURL.absoluteString)
+    
+    if let data = data,
+        let favicon = try? JSONDecoder().decode(Favicon.self, from: data) {
+      return favicon
+    }
+    
+    return nil
```
Comment (kylehickinson) [inline BraveShared/Favicons/FaviconFetcher.swift]: 1. Same comment as above, don't do optional try, catch error and log
2. Should rearrange to return nil early

File: BraveShared/Favicons/FaviconRenderer.swift [hunk 19]
```diff
@@ -0,0 +1,87 @@
...
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+import Foundation
+import UIKit
+import BraveCore
+
+/// A class for rendering a FavIcon onto a `UIImage`
+class FaviconRenderer {
```
Comment (kylehickinson) [inline BraveShared/Favicons/FaviconRenderer.swift:11]: You have multiple classes that do the same exact setup and execution (`BundledFaviconImageRenderer`, `FaviconRenderer`), could this not be simplified or consolidated?

File: BraveShared/Favicons/UIImage+FaviconRenderer.swift [hunk 20]
```diff
@@ -0,0 +1,209 @@
...
+
+    let text = (monogramString ?? FaviconUtils.monogramLetter(
+        for: url,
+        fallbackCharacter: nil
+      )) as NSString
+
+    let padding = 4.0
+    let finalImage = drawOnImageContext { context, rect in
+      guard let font = label.font else { return }
+      var fontSize = font.pointSize
```
Comment (kylehickinson) [inline BraveShared/Favicons/UIImage+FaviconRenderer.swift:154]: How will this work with dynamic font? `UILabel`'s font by default is going to be the preferred body font which means it will grow as the users font size grows. Have you tested the monogram code with multiple font sizes?

File: BraveShared/Favicons/UIImage+FaviconRenderer.swift [hunk 21]
```diff
@@ -0,0 +1,209 @@
...
+
+// MARK: - Drawing
+extension UIImage {
+  private static func drawOnImageContext(size: CGSize, _ draw: (CGContext, CGRect) -> Void) -> UIImage? {
+    let size = CGSize(width: size.width, height: size.height)
+    UIGraphicsBeginImageContextWithOptions(size, false, UIScreen.main.scale)
+    draw(UIGraphicsGetCurrentContext()!, CGRect(size: size))
+    let img = UIGraphicsGetImageFromCurrentImageContext()
+    UIGraphicsEndImageContext()
+    return img
```
Comment (kylehickinson) [inline BraveShared/Favicons/UIImage+FaviconRenderer.swift]: You should be using the newer `UIGraphicsImageRenderer` API

File: BraveWallet/FaviconReader.swift [hunk 22]
```diff
@@ -1,31 +1,11 @@
 import SwiftUI
-
-public protocol WalletFaviconRenderer {
-  func loadIcon(siteURL: URL, persistent: Bool, completion: ((UIImage?) -> Void)?)
-}
```
Comment (kylehickinson) [inline BraveWallet/FaviconReader.swift:5]: cc @nuo-xu @StephenHeaps re: these changes

File: Client/Frontend/Browser/BrowserViewController/BVC+Rewards.swift [hunk 23]
```diff
@@ -304,13 +304,9 @@ extension Tab {
...
 
     group.notify(queue: .main) {
-      let faviconURL = URL(string: self.displayFavicon?.url ?? "")
-      if faviconURL == nil {
-        log.warning("No favicon found in \(self) to report to rewards panel")
-      }
       rewards.reportLoadedPage(
         url: url, redirectionURLs: urls.isEmpty ? [url] : urls,
-        faviconURL: faviconURL, tabId: Int(self.rewardsId),
+        faviconURL: nil, tabId: Int(self.rewardsId),
```
Comment (kylehickinson) [inline Client/Frontend/Browser/BrowserViewController/BVC+Rewards.swift:309]: Why is this nil'd out?

File: Client/Frontend/Browser/FaviconHandler.swift [hunk 24]
```diff
@@ -25,82 +25,23 @@ class FaviconHandler {
...
-        guard let image = image,
-              let imageData = data
-        else {
-          favicon.width = 0
-          favicon.height = 0
-          
-          onSuccess(favicon, data)
-          return
+        if let favicon = favicon {
+          tab.favicons.append(favicon)
```
Comment (kylehickinson) [inline Client/Frontend/Browser/FaviconHandler.swift]: This shouldn't occur in the continuation like this, you should get the result and add it after like:

```swift
let favicon = try await withCheckedThrowingContinuation { ... }
tab.favicons.append(favicon)
return favicon
```

File: Client/Frontend/Browser/FaviconHandler.swift [hunk 25]
```diff
@@ -110,18 +51,18 @@ extension FaviconHandler: TabEventHandler {
     tab.favicons.removeAll(keepingCapacity: false)
     Task { @MainActor in
       if let iconURL = metadata.largeIconURL {
-        let (favicon, data) = try await loadFaviconURL(iconURL, type: .appleIcon, forTab: tab)
-        TabEvent.post(.didLoadFavicon(favicon, with: data), for: tab)
+        let favicon = try await loadFaviconURL(iconURL, forTab: tab)
+        TabEvent.post(.didLoadFavicon(favicon, with: nil), for: tab)
```
Comment (kylehickinson) [inline Client/Frontend/Browser/FaviconHandler.swift]: Are there even any places you'd call this with data anymore? If not then we should just remove it from the TabEvent

File: Client/Frontend/Browser/Tab.swift [hunk 26]
```diff
@@ -503,7 +505,7 @@ class Tab: NSObject {
 
   var displayFavicon: Favicon? {
     if let url = url, InternalURL(url)?.isAboutHomeURL == true { return nil }
-    return favicons.max { $0.width! < $1.width! }
+    return favicons.first
```
Comment (kylehickinson) [inline Client/Frontend/Browser/Tab.swift:509]: Is the first one guaranteed to be the highest sized one now?

File: Client/Frontend/Browser/Tab.swift [hunk 27]
```diff
@@ -108,6 +108,8 @@ class Tab: NSObject {
 
   var userActivity: NSUserActivity?
 
+  let faviconDriver = FaviconLoader.Driver(privateBrowsingMode: false)
```
Comment (kylehickinson) [inline Client/Frontend/Browser/Tab.swift]: Hardcoded private browsing to false? This should be set when the tab is being configured and you know whether or not its private

File: Client/Frontend/Browser/TabManager.swift [hunk 28]
```diff
@@ -313,19 +313,11 @@ class TabManager: NSObject {
     guard let newSelectedTab = tab, let previousTab = previous, let newTabUrl = newSelectedTab.url, let previousTabUrl = previousTab.url else { return }
 
     if !PrivateBrowsingManager.shared.isPrivateBrowsing {
-      let previousFaviconURL = URL(string: previousTab.displayFavicon?.url ?? "")
-      if previousFaviconURL == nil && !previousTabUrl.isLocal {
-        rewardsLog.warning("No favicon found in \(previousTab) to report to rewards panel")
-      }
       rewards?.reportTabUpdated(
-        Int(previousTab.rewardsId), url: previousTabUrl, faviconURL: previousFaviconURL, isSelected: false,
+        Int(previousTab.rewardsId), url: previousTabUrl, faviconURL: nil, isSelected: false,
```
Comment (kylehickinson) [inline Client/Frontend/Browser/TabManager.swift:317]: Why are these nil'd out?

File: Client/Frontend/Browser/UserScriptManager.swift [hunk 29]
```diff
@@ -435,6 +435,31 @@ class UserScriptManager {
...
+
+    replacements.forEach({
+      alteredSource = alteredSource.replacingOccurrences(of: $0.key, with: $0.value, options: .literal)
+    })
+
+    return WKUserScript.create(
+      source: alteredSource,
+      injectionTime: .atDocumentEnd,
+      forMainFrameOnly: true,
+      in: .page)
```
Comment (kylehickinson) [inline Client/Frontend/Browser/UserScriptManager.swift]: Why can't this be sandboxed? We aren't hooking anything I thought?

**Review by kylehickinson — CHANGES_REQUESTED** [review]
1. Lets move everything favicon related to its own `Favicons` module instead of shoving it in BraveShared and Client
2. Why did we choose to use `DispatchWorkItem` everywhere instead of Swift concurrency? 
3. Has this been tested with the favourites widget? I notice we now have 2 types of favicons, but not sure how this affects that
4. Has this been tested with multiple font sizes

File: BraveShared/Extensions/FileManagerExtension.swift (hunk 14 above)
Comment (Brandon-T) [reply BraveShared/Extensions/FileManagerExtension.swift:20]: It is in `BraveShared` -> `BraveShared/Extensions/FileManagerExtension.swift`

File: BraveShared/Favicons/UIImage+FaviconRenderer.swift (hunk 20 above)
Comment (Brandon-T) [reply BraveShared/Favicons/UIImage+FaviconRenderer.swift:154]: It shouldn't be. It's an image. The image will be scaled if the UI scales. Chromium doesn't handle it either so if we used the `monogram` image that Chromium gives us, it would still not scale the font because the font itself is applied to the image, then the image is scaled with the UI.

File: Client/Frontend/Browser/BrowserViewController/BVC+Rewards.swift (hunk 23 above)
Comment (Brandon-T) [reply Client/Frontend/Browser/BrowserViewController/BVC+Rewards.swift:309]: Because we don't have a `faviconURL`. The URL for the favicon is pulled from Brave-Core now.

File: Sources/Shared/UIColorExtensions.swift (hunk 13 above)
Comment (Brandon-T) [reply Sources/Shared/UIColorExtensions.swift:45]: If I move my favicon code to `BraveFavicons` module, it will have to import `BraveUI` which will do:
`BraveShared -> BraveFavicons -> BraveUI -> BraveShared`

Widgets are all in `BraveShared` so if I move favicons to its own module, it will import BraveShared for FileManager, etc.. and BraveShared would import it for Widget icons.

File: BraveShared/Extensions/FileManagerExtension.swift (hunk 14 above)
Comment (kylehickinson) [reply BraveShared/Extensions/FileManagerExtension.swift:20]: Sorry, thought I read the file name was in Client for some reason

File: Client/Frontend/Browser/BrowserViewController/BVC+Rewards.swift (hunk 23 above)
Comment (kylehickinson) [reply Client/Frontend/Browser/BrowserViewController/BVC+Rewards.swift:309]: We may need to check if this affects the rewards panel though because we show a favicon there for the publisher... Please visit `wikipedia.org` and ensure it displays the favicon like so:
image

File: Sources/Shared/UIColorExtensions.swift (hunk 13 above)
Comment (kylehickinson) [reply Sources/Shared/UIColorExtensions.swift:45]: 1. We don't need to prefix it, it can just be named `Favicons` (Everything in the package is for Brave, BraveShared + BraveUI are legacy names)
2. The widgets stuff should use the new favicons stuff, no? We don't want them to fall out of sync. See my other comment about the 2 favicon types below
3. If we have a cycle then we want to correct the APIs bounds. `BraveShared` is legacy, we can break things out of it and create more modules as needed (i.e. make a `Widgets` module or something).  We should be looking to break up BraveShared + BraveUI into smaller more concise modules, not just adding more to them because they're there.  What widgets stuff aside from the duped.  We can also look into removing the BraveShared dependency in BraveUI

File: BraveShared/Favicons/FaviconRenderer.swift (hunk 19 above)
Comment (Brandon-T) [reply BraveShared/Favicons/FaviconRenderer.swift:11]: They do different things. 
One class is for rendering Favicons from Brave-Core only (including monogram images).
The other class is for loading and rendering Favicons that are Bundled with the app (TopSites), no monograms at all.

File: Client/Frontend/Browser/BrowserViewController/BVC+Rewards.swift (hunk 23 above)
Comment (Brandon-T) [reply Client/Frontend/Browser/BrowserViewController/BVC+Rewards.swift:309]: It works correctly in the rewards panel. This is just for some reporting thing.

File: Client/Frontend/Browser/Tab.swift (hunk 26 above)
Comment (Brandon-T) [reply Client/Frontend/Browser/Tab.swift:509]: Yeah. When we pass `.desiredMedium` we're asking Brave-Core to give us an icon <= `192x192` whichever is closest. There could be icons that are 256x256, but they are ignored atm until my other Brave-Core PR is merged.
BraveCore only returns `one` icon anyway. It's just that our original code has `favicons` array in `Tab` as if it was loading `multiple`, even though it only ever used one? So I kept it that way just in case we do need multiple.

File: Sources/Shared/UIColorExtensions.swift (hunk 13 above)
Comment (Brandon-T) [reply Sources/Shared/UIColorExtensions.swift:45]: Widgets are already using the latest Favicon code in this PR. But this PR shouldn't be for refactoring the whole app.
I've only added one variable `var rgb: Int` to the Brave-Shared code. Refactoring all that stuff for that isn't worth it for me.

That should be another separate PR to break the chains of all these dependencies. It's hundreds more changes for me to move Widgets to its own module, then move Favicons to its own, then break the links between all these other modules.

File: BraveShared/Favicons/FaviconRenderer.swift (hunk 19 above)
Comment (kylehickinson) [reply BraveShared/Favicons/FaviconRenderer.swift:11]: I know they do different things, I said that they have the exact same setup + execution, in fact the two files are mostly duplicated aside from the body of the task work and the name of the load method (which could be the same).

File: Sources/Shared/UIColorExtensions.swift (hunk 13 above)
Comment (kylehickinson) [reply Sources/Shared/UIColorExtensions.swift:45]: You are already changing 261 files in this PR, so that train has left the station 🤷‍♂️ Also a large portion of this is moving files into BraveShared, so moving it into their own module isn't far off. 

Both of these things though are still not the root of this comment which is that we shouldn't be adding new code to `Shared` anymore. Anyways, you probably don't even need this anymore once you deleted `FaviconAttributesContainer`

File: Client/Frontend/Browser/BrowserViewController/BVC+Rewards.swift (hunk 23 above)
Comment (kylehickinson) [reply Client/Frontend/Browser/BrowserViewController/BVC+Rewards.swift:309]: I think this will probably break some underlying code in the rewards system but it may not matter in this case anymore since we don't show any of the details

File: BraveShared/Favicons/FaviconRenderer.swift (hunk 19 above)
Comment (Brandon-T) [reply BraveShared/Favicons/FaviconRenderer.swift:11]: So I made the method names the same. I don't see what other setup or execution is the same. 
They are 129 out of 169 total changes, difference between them. If I make the methods the same signature, and the class names the same: 127 of 169 for a 75% difference between the two classes :S
The other one has a bunch of static variables that aren't necessary for the BraveCore one.

I don't see any reason to merge these classes to get the big class we had before.

File: package-lock.json [hunk 30]
```diff
@@ -295,9 +295,9 @@
       }
     },
     "node_modules/brave-core-ios": {
-      "version": "1.41.91",
-      "resolved": "https://github.com/brave/brave-browser/releases/download/v1.41.91/brave-core-ios-1.41.91.tgz",
-      "integrity": "sha512-zi8CP7pN3ftD1QikjsiQbGfU50rK0J9Mco2vz2EdUBkkkHDWVAaduZ5uDrTSziLbVD1DXB/5xXXAbz3yWHsdAw==",
+      "version": "1.41.82",
+      "resolved": "https://github.com/brave/brave-browser/releases/download/v1.41.82/brave-core-ios-1.41.82.tgz",
+      "integrity": "sha512-1yKuW2hf7Pk8uFDyd3o+wOWQiQBkYU7am0JWuHGHohDtQN5TYBTv1YQI1NnEYBDrJQOZigo5fnEkpLJtUT1XMw==",
```
Comment (kylehickinson) [inline package-lock.json]: This package-lock file change needs to be reverted/removed from the PR

**Review by kylehickinson — CHANGES_REQUESTED** [review]

File: package-lock.json (hunk 30 above)
Comment (Brandon-T) [reply package-lock.json]: Reverted/rewrote the commit.

File: Sources/Shared/UIColorExtensions.swift (hunk 13 above)
Comment (Brandon-T) [reply Sources/Shared/UIColorExtensions.swift:45]: It is needed for `try UIColor(rgb: container.decode(Int.self, forKey: .backgroundColor))` in `Favicon.swift`
I can move the extension to `BraveShared` instead of `Shared`.
